### PR TITLE
Add leash effect and proper fixes.

### PR DESCRIPTION
**Pull request by TheLimeGlass**
Adds the ability to leash entities.
And needed fixes

### CODE REVIEW

File: src/main/java/ch/njol/skript/effects/EffContinue.java [hunk 1]
```diff
@@ -70,6 +70,7 @@ public String toString(@Nullable Event e, boolean debug) {
 		return "continue";
 	}
 
+	@SuppressWarnings("null")
 	@Override
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffContinue.java]: why did you add this

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import ch.njol.util.Kleenean;
+
+@Name("Leash entities")
+@Description("Leash living entities to other entities")
+@Examples("leash the player to target entity")
+@Since("INSERT VERSION")
+public class EffLeash extends Effect {
+
+	static {
+		Skript.registerEffect(EffLeash.class, "(leash|lead) %livingentities% to %entities%", "make %entities% (leash|lead) %livingentities%");
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: put the patterns on different lines

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		for (Entity holder : holders.getArray(event)) {
+			for (LivingEntity target : targets.getArray(event)) {
+				target.setLeashHolder(holder);
+			}
+		}
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "leash " + targets.toString(event, debug) + " to " + holders.toString(event, debug);
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: use `e` like everywhere else

File: src/main/java/ch/njol/skript/effects/EffToggleFlight.java [hunk 4]
```diff
@@ -50,7 +50,7 @@
 
 	private boolean allow;
 
-	@SuppressWarnings("unchecked")
+	@SuppressWarnings({"unchecked", "null"})
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffToggleFlight.java]: why is this here

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		holders = (Expression<Entity>) expressions[1 - matchedPattern];
+		targets = (Expression<LivingEntity>) expressions[matchedPattern];
+		return true;
+	}
+	
+	@Override
+	protected void execute(Event event) {
+		for (Entity holder : holders.getArray(event)) {
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: what's the point in allowing multiple holders if they are just immediately reset anyway

File: src/main/java/ch/njol/skript/effects/EffContinue.java (hunk 1 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffContinue.java]: Because there is a warning without it. Use the proper IDE to see.

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 2 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: It doesn't exceed the maximum characters and looks better.

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 5 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: What do you mean reset?

File: src/main/java/ch/njol/skript/effects/EffToggleFlight.java (hunk 4 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffToggleFlight.java]: It creates a warning, use the proper IDE to view.

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 5 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: the only leash that takes effect is the last looped thing

File: src/main/java/ch/njol/skript/effects/EffContinue.java (hunk 1 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/effects/EffContinue.java]: there's warnings all over skript. The class was already fine to ECJ.

File: src/main/java/ch/njol/skript/effects/EffToggleFlight.java (hunk 4 above)
Comment (Snow-Pyon) [reply src/main/java/ch/njol/skript/effects/EffToggleFlight.java]: There is no "proper" IDE, even IntelliJ can warn about these things if configured that way, so there is no need to debate about something that's completely personal such as IDE preference.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffToggleFlight.java]: Then it should have been seen as these were new additions.

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 5 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: I'm confused... You know an entity can have multiple leashes to it right?

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	}
+	
+	@SuppressWarnings("null")
+	private Expression<Entity> holders;
+	@SuppressWarnings("null")
+	private Expression<LivingEntity> targets;
+
+	@SuppressWarnings({"unchecked", "null"})
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: expressions --> exprs

Please don't change standarts in only one thing.

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 5 above)
Comment (Blueyescat) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: But a target (not holder) can't have multiple leashes.

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 6 above)
Comment (jaylawl) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: This should be an expression rather than an effect

**Comment by jaylawl** [comment]
This should be an expression rather than an effect

**Comment by Pikachu920** [comment]
nothing wrong with it being an effect, but having it as an expression could be useful too

**Comment by Blueyescat** [comment]
If there will be an expression it shouldn't be changeable like on player visibility stuff. @Snow-Pyon suggested this to me and makes sense IMO.

**Comment by jaylawl** [comment]
I don't get what's wrong with using a property expression instead. that property would combine the effect, conditions and an unleash effect too...

```
if entity's leash holder is set
set {_livingentity} to entity's leash holder
set entity's leash holder to player
reset entity's leash holder
```

as opposed to having 2 effects + 1 conditions (and still be lacking to `get` method)

**Comment by TheBentoBox** [comment]
An entity can have multiple leash holders and I don't believe a simple property expression can handle multiple values which would be a problem there.

**Comment by jaylawl** [comment]
@TheBentoBox it can be attached to multiple leashs - are you sure? Where is that information from?

**Comment by TheBentoBox** [comment]
I'm just going off of experience, not like I'm sourcing it, but I may be misremembering as well. If they can't have multiple leash holders (or if the Bukkit API doesn't support it, which may be the case because I do think I remember the leash holder methods dealing with singular entities) then I agree a property expression is an easy solution. I don't love `delete entity's leash holder` being how you unleash it though. That, to me, sounds like it would delete the actual entity which is the target's leash holder.

**Comment by jaylawl** [comment]
The API is definitly only allowing for a single entity to be the leash holder: https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/LivingEntity.html#getLeashHolder--
I agree, `delete leash holder` doesn't sound good, and neither really does `reset leash holder` :|

**Comment by Blueyescat** [comment]
@jaylawl Because effects are more english.

An entity can't have multiple leash holders but a leash holder can leash multiple leashed entities.

**Comment by Nicofisi** [comment]
Also @TheLimeGlass next time instead of commenting 'last commit was to fix the license header.' please just properly name the commit, we'll probably squash the whole thing before merging anyway, but proper commit titles make it way more intuitive to review on GitHub

File: src/main/java/ch/njol/skript/effects/EffContinue.java (hunk 1 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffContinue.java]: No it wasn't

**Comment by TheLimeGlass** [comment]
@Nicofisi commit history

**Comment by Pikachu920** [comment]
@jaylawl effects can be more natural, take flight for example - what's more natural?
This:
```
set player's flight mode to true
```
or this:
```
allow player to fly
```

**Comment by TheLimeGlass** [comment]
@jaylawl property expressions are for single types contained in an expression, for example name of %player% there is only one value and the object that is being got is a string of their name, in this effect there needs to be two expressions (called a COMBINED) and doesn't need to be an expression because having the expression would be tricky and un-needed for setting. You would have some patterns that would only be used for setting due to this. So keeping it as an effect is needed.

**Comment by jaylawl** [comment]
Well in that case, please disregard any prior "protest" of mine

File: src/main/java/ch/njol/skript/effects/EffContinue.java (hunk 1 above)
Comment (Nicofisi) [reply src/main/java/ch/njol/skript/effects/EffContinue.java]: @TheLimeGlass I understand that you wanted good and making a separate PR would take too much effort/be pointless for only this change and I kinda feel you and agree (kinda because I wouldn't change this at all, this should be fixed along with the other 400 warnings IMO lol), but at least make a separate commit next time please. ~~It's easy and if it isn't then switch to IntelliJ because it is there~~

**Comment by TheLimeGlass** [comment]
I always do fixes along with my pull requests, quality will just down spiral.

File: src/main/java/ch/njol/skript/effects/EffContinue.java (hunk 1 above)
Comment (Nicofisi) [reply src/main/java/ch/njol/skript/effects/EffContinue.java]: That's.. not exactly.. what I meant

**Comment by Blueyescat** [comment]
unleash doesn't need a leash holder.

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+		if (entity == null && !unleash)
+			return;
+		for (LivingEntity target : targets.getArray(e)) {
+			target.setLeashHolder(!unleash ? entity : null);
+		}
+	}
+
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
+		return unleash ? "un" : "" + "leash " + targets.toString(e, debug) + (unleash ? " from " : "" + " to ") + holder.toString(e, debug);
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: This gets translated by Java to
```java
return unleash ? "un" : ("" + "leash " + targets.toString(e, debug) + (unleash ? " from " : "" + " to ") + holder.toString(e, debug));
```
what you need is
```java
return (unleash ? "unleash " : "leash ") + targets.toString(e, debug) + (unleash ? "" : " to " + holder.toString(e, debug));
```
or maybe something even cleaner

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	@SuppressWarnings("null")
+	private Expression<Entity> holder;
+	@SuppressWarnings("null")
+	private Expression<LivingEntity> targets;
+	private boolean unleash;
+
+	@SuppressWarnings({"unchecked", "null"})
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		holder = (Expression<Entity>) exprs[1 - matchedPattern];
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: ArrayOutOfBoundException for unleash

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		targets = (Expression<LivingEntity>) exprs[matchedPattern];
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: ArrayOutOfBoundException for unleash

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+		unleash = matchedPattern == 2;
+		return true;
+	}
+	
+	@Override
+	protected void execute(Event e) {
+		Entity entity = holder.getSingle(e);
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: `Entity holder = this.holder.getSingle(e);` would be a better naming imo

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+		Skript.registerEffect(EffLeash.class,
+				"(leash|lead) %livingentities% to %entity%",
+				"make %entity% (leash|lead) %livingentities%",
+				"un(leash|lead) [holder of] %livingentities%");
+	}
+	
+	@SuppressWarnings("null")
+	private Expression<Entity> holder;
+	@SuppressWarnings("null")
+	private Expression<LivingEntity> targets;
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: Since you added unleash this should be @Nullable

File: src/main/java/ch/njol/skript/conditions/CondLeashHolder.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+@Name("Has Leash Holder")
+@Description("Checks whether entities have leash holders.")
+@Examples("target entity doesn't have a leash holder")
+@Since("INSERT VERSION")
+public class CondLeashHolder extends Condition {
+	
+	static {
+		Skript.registerCondition(CondLeashHolder.class,
+				"%livingentities% (has|do[es] [have]|contains) [a] leash holder [from %-entity%]",
+				"%livingentities% (has([n't]| not [got])|do[es](n't| not) (contain|have)) [a] leash holder [from %-entity%]");
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/conditions/CondLeashHolder.java]: image

Comment (Nicofisi) [reply src/main/java/ch/njol/skript/conditions/CondLeashHolder.java]: `event-entity contains leash holder from player` what?

Comment (Blueyescat) [reply src/main/java/ch/njol/skript/conditions/CondLeashHolder.java]: What are these patterns lol they are really bad. 
**`(do[es]n't|don't|do[es] not)`** is the currently used pattern and better imo.
And **`ha(s|ve)`**.

contains is also so absurd.

Comment (Blueyescat) [reply src/main/java/ch/njol/skript/conditions/CondLeashHolder.java]: Also make the class name CondHasLeash like others..

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 8 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: 0, 1, 2? There are three patterns is there not...

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 9 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: 0, 1, 2? There are three patterns is there not...

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 11 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: a boolean can't be null, it's default value is always false.

File: src/main/java/ch/njol/skript/conditions/CondLeashHolder.java (hunk 12 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/conditions/CondLeashHolder.java]: These are worse

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 13]
```diff
@@ -64,17 +64,17 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
...
 		for (LivingEntity target : targets.getArray(e)) {
-			target.setLeashHolder(!unleash ? entity : null);
+			target.setLeashHolder(!unleash ? holder : null);
 		}
 	}
 
 	@Override
 	public String toString(@Nullable Event e, boolean debug) {
-		return unleash ? "un" : "" + "leash " + targets.toString(e, debug) + (unleash ? " from " : "" + " to ") + holder.toString(e, debug);
+		return unleash ? "un" : "" + "leash " + targets.toString(e, debug) + (unleash ? "" : " to ") + holder.toString(e, debug);
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: Still missing braces

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 9 above)
Comment (Nicofisi) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: where is there a third expression in the pattern

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 11 above)
Comment (Nicofisi) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: this comment is about the line above

File: src/main/java/ch/njol/skript/conditions/CondStartsWith.java [hunk 14]
```diff
@@ -50,11 +50,11 @@
 	private Expression<String> strings;
 	@SuppressWarnings("null")
 	private Expression<String> prefix;
-	private boolean ends;
+	boolean ends;
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/conditions/CondStartsWith.java]: what the hell are you changing all these to package-private

File: src/main/java/ch/njol/skript/effects/EffLeash.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+		if (holder == null && !unleash)
+			return;
+		for (LivingEntity target : targets.getArray(e)) {
+			target.setLeashHolder(!unleash ? holder : null);
+		}
+	}
+
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
+		return unleash ? "un" : "" + "leash " + targets.toString(e, debug) + (unleash ? "" : " to ") + holder.toString(e, debug);
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/effects/EffLeash.java]: you still are missing a pair of brackets here

File: src/main/java/ch/njol/skript/conditions/CondStartsWith.java (hunk 14 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/conditions/CondStartsWith.java]: I stated above

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 13 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: What?

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 9 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: Matched pattern does not return how many expressions there are... It returns 0-2 if there are three different syntax. Now 4 so if it's 0-1 it's negated is true and if it's 2-3 it's false. Basic Skript.

Comment (Nicofisi) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: are we talking about the same line? the comments are always to the line above

`exprs[matchedPattern];` if you use the third patternnit will call exprs[2] and there is only one expression in this pattern

"basic skript"

File: src/main/java/ch/njol/skript/effects/EffLeash.java (hunk 13 above)
Comment (Blueyescat) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: (unleash ? "un" : "" + "leash ")

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: What about it?

Comment (Nicofisi) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: Exactly, you need then but you dont have them here, thats what I've meant since the beginning lol

`return unleash ? "un" : "" + "leash " + targets.......` - your code

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLeash.java]: This doesn't exist bro...

File: src/main/java/ch/njol/skript/conditions/CondStartsWith.java (hunk 14 above)
Comment (Snow-Pyon) [reply src/main/java/ch/njol/skript/conditions/CondStartsWith.java]: In simple words, it is to avoid warnings about synthetic-access since I am also accessing that variable in an inner class.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/conditions/CondStartsWith.java]: ^

**Comment by bensku** [comment]
Manually merged. Naturally, I ignored changes to unrelated files; they would've caused merge conflicts.
